dao-voting/src/main.sw

```diff
@@ -0,0 +1,312 @@
+
+        let sender: Identity = msg_sender().unwrap();
+        let sender_balance = storage.balances.get(sender);
+
+        require(vote_amount <= sender_balance, UserError::NotEnoughAssets);
+
+        if (is_yes_vote) {
+            proposal.yes_votes += vote_amount;
+        } else {
+            proposal.no_votes += vote_amount;
+        };
+
+        storage.balances.insert(sender, sender_balance - vote_amount);
+
+        storage.votes.insert((sender, proposal_id), storage.votes.get((sender, proposal_id)) + vote_amount);
+
+        storage.proposals.insert(proposal_id, proposal);
+
+        log(VoteEvent {
+            id: proposal_id, user: sender, vote_amount: vote_amount
+        });
+    }
+
+    /// Execute a given proposal
+    ///
+    /// # Arguments
+    ///
+    /// - `proposal_id` - proposal to execute
+    ///
+    /// # Reverts
+    ///
+    /// * When the construct has not been called to initialize
+    /// * When the proposal id is out of range
+    /// * When the proposal has not expired
+    /// * When the proposal has not met the necessary approval percentage
+    #[storage(read, write)]fn execute(proposal_id: u64) {
+        require(proposal_id < storage.proposal_count, UserError::InvalidId);
+
+        let proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height < height(), ProposalError::ProposalStillActive);
+
+        // Prevents divide by 0 error when calculating acceptance_percentage below
+        if (proposal.yes_votes == 0 && proposal.no_votes == 0) {
+            revert(42);
+        }
+
+        // TODO figure out how to prevent approval percentage from overflowing
+        // When close to the u64 max
+        // https://github.com/FuelLabs/sway-applications/issues/106
+        let acceptance_percentage = proposal.yes_votes * 100 / (proposal.yes_votes + proposal.no_votes);
+        require(proposal.acceptance_percentage <= acceptance_percentage, ProposalError::ApprovalPercentageNotMet);
+
+        asm(memory_address: proposal.call_data.memory_address, num_coins_to_forward: proposal.call_data.num_coins_to_forward, asset_id_of_coins_to_forward: proposal.call_data.asset_id_of_coins_to_forward, amount_of_gas_to_forward: proposal.call_data.amount_of_gas_to_forward) {
+            call memory_address num_coins_to_forward asset_id_of_coins_to_forward amount_of_gas_to_forward;
+        }
+        // Users can now convert their votes back into tokens
+        log(ExecuteEvent {
+            id: proposal_id, 
+        });
+    }
+
+    /// Unlock governance tokens which have been locked by users who have voted on a proposal
+    /// If the user had not voted on the given, expired proposal then nothing happens
+    ///
+    /// # Arguments
+    ///
+    /// - `proposal_id` - proposal to turn user votes back into governance tokens
+    ///
+    /// # Reverts
+    ///
+    /// * When the constructor has not ben called to initialize
```

Comment:
Man this constructor is really giving you a hard time. If only the constructor wasn't mean enough to point it out each time :)